core/foundation/inc/ROOT/RRangeCast.hxx

```diff
@@ -0,0 +1,207 @@
+      return static_cast<T>(u);
+   }
+};
+
+#endif
+
+// For SFINAE-based checks for the existence of the `begin` and `end` methods.
+template <typename T>
+constexpr auto hasBeginEnd(int) -> decltype(std::begin(std::declval<T>()), std::end(std::declval<T>()), true)
+{
+   return true;
+}
+
+template <typename>
+constexpr bool hasBeginEnd(...)
+{
+   return false;
+}
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
+
+public:
+   TypedIter(WrappedIterator_t const &iter) : fIter{iter} {}
+
+   TypedIter &operator++()
+   {
+      ++fIter;
+      return *this;
+   }
+   TypedIter operator++(int)
+   {
+      TypedIter tmp(*this);
+      operator++();
+      return tmp;
+   }
+   bool operator==(const TypedIter &rhs) const { return fIter == rhs.fIter; }
+   bool operator!=(const TypedIter &rhs) const { return fIter != rhs.fIter; }
+
+   void swap(TypedIter &other) { fIter.swap(other.fIter); }
+
+   // We want to know at compile time whether dynamic_cast or static_cast is
+   // used. First of all to avoid overhead, but also to avoid a compiler
+   // error when using dynamic_cast on a non-polymorphic class. In C++17,
+   // this can be done easily with `if constexpr`, but for the older
+   // standards we have to use a more verbose alternative. Both ways are
+   // explicitely implemented for different standards, so that when the
+   // minimum C++ standard for ROOT is raised to C++17 it's easy to remember
+   // that we can avoid much boilerplate code in this file.
+#if (__cplusplus < 201700L)
+   T operator*() { return ROOT::Internal::RCast<T, isDynamic>::cast(*fIter); }
+#else
+   T operator*()
+   {
+      if constexpr (isDynamic) {
+         return dynamic_cast<T>(*fIter);
+      } else {
+         if constexpr (std::is_polymorphic<RBaseType<T>>::value) {
+            assert(dynamic_cast<T>(*fIter));
+         }
+         return static_cast<T>(*fIter);
+      }
+   }
+#endif
+
+private:
+   WrappedIterator_t fIter;
+};
+
+} // namespace Internal
+
+/// Wraps any collection that can be used in range-based loops and applies
+/// `static_cast<T>` or `dynamic_cast<T>` to each element.
+template <typename T, bool isDynamic, typename Range_t>
+class RRangeCast {
+
+public:
+   RRangeCast(Range_t &&inputRange) : fInputRange{inputRange}
+   {
+      static_assert(ROOT::Internal::hasBeginEnd<Range_t>(0),
+                    "Type with no `begin` or `end` method passed to `RRangeCast`");
+   }
+
+   using const_iterator = Internal::TypedIter<T, decltype(std::cbegin(std::declval<Range_t>())), isDynamic>;
+   const_iterator begin() const { return std::cbegin(fInputRange); }
+   const_iterator end() const { return std::cend(fInputRange); }
+
+   using iterator = Internal::TypedIter<T, decltype(std::begin(std::declval<Range_t>())), isDynamic>;
+   iterator begin() { return std::begin(fInputRange); }
+   iterator end() { return std::end(fInputRange); }
+
+private:
+   Range_t fInputRange;
```

Comment:
Are you creating copies of containers here when called in the wrong way?

Reply:
Sure, if you call it in the wrong way and instantiate a `RRangeCast` with `Range_t` as a value type  explicitly, you can cause unnecessary moving (not copying, as the constructor has the signature `RRangeCast(Range_t &&inputRange)`).

However, if you use the `RangeStaticCast` or `RangeDynCast` functions that take a universal reference that is perfect-forwarded to `RRangeCast`, `Range_t` will always be a reference type.